pinot-core/src/test/java/org/apache/pinot/core/query/reduce/BrokerReduceServiceTest.java

```diff
@@ -0,0 +1,79 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.reduce;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.common.exception.QueryException;
+import org.apache.pinot.common.request.BrokerRequest;
+import org.apache.pinot.common.response.broker.BrokerResponseNative;
+import org.apache.pinot.common.response.broker.QueryProcessingException;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.common.utils.DataSchema.ColumnDataType;
+import org.apache.pinot.common.utils.DataTable;
+import org.apache.pinot.core.common.datatable.DataTableBuilder;
+import org.apache.pinot.core.transport.ServerRoutingInstance;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.sql.parsers.CalciteSqlCompiler;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+
+
+public class BrokerReduceServiceTest {
+  private static final CalciteSqlCompiler COMPILER = new CalciteSqlCompiler();
+
+  @Test
+  public void testReduceTimeout()
+      throws IOException {
+    Map<String, Object> properties = new HashMap<>();
+    properties.put(CommonConstants.Broker.CONFIG_OF_MAX_REDUCE_THREADS_PER_QUERY, 2);
+    BrokerReduceService brokerReduceService = new BrokerReduceService(new PinotConfiguration(properties));
+
+    BrokerRequest brokerRequest = COMPILER.compileToBrokerRequest(
+        "SELECT COUNT(*) FROM testTable GROUP BY col1 OPTION(groupByMode=sql,responseFormat=sql)");
+    DataSchema dataSchema =
+        new DataSchema(new String[]{"col1", "count(*)"}, new ColumnDataType[]{ColumnDataType.INT, ColumnDataType.LONG});
+    DataTableBuilder dataTableBuilder = new DataTableBuilder(dataSchema);
+    int numGroups = 5000;
+    for (int i = 0; i < numGroups; i++) {
+      dataTableBuilder.startRow();
+      dataTableBuilder.setColumn(0, i);
+      dataTableBuilder.setColumn(1, 1L);
+      dataTableBuilder.finishRow();
+    }
+    DataTable dataTable = dataTableBuilder.build();
+    Map<ServerRoutingInstance, DataTable> dataTableMap = new HashMap<>();
+    int numInstances = 1000;
+    for (int i = 0; i < numInstances; i++) {
+      ServerRoutingInstance instance = new ServerRoutingInstance("localhost", i, TableType.OFFLINE);
+      dataTableMap.put(instance, dataTable);
+    }
+    long reduceTimeoutMs = 1;
+    BrokerResponseNative brokerResponse =
+        brokerReduceService.reduceOnDataTable(brokerRequest, brokerRequest, dataTableMap, reduceTimeoutMs, null);
```

Comment:
brokerRequest param is duplicated.

Reply:
This is intentional. The first arg is the broker side broker request, second arg is the server side broker request.